File: hu/codes/c/chapter_dynamic_programming/climbing_stairs_backtrack.c
/**
 * File: climbing_stairs_backtrack.c
 * Created Time: 2023-09-22
 * Author: huawuque404 (huawuque404@163.com)
 */

#include "../utils/common.h"

/* Visszalépés */
void backtrack(int *choices, int state, int n, int *res, int len) {
    // Ha elértük az n-edik lépcsőfokot, növeljük a megoldások számát 1-gyel
    if (state == n)
        res[0]++;
    // Az összes lehetőség bejárása
    for (int i = 0; i < len; i++) {
        int choice = choices[i];
        // Vágás: nem szabad az n-edik lépcsőfokot meghaladni
        if (state + choice > n)
            continue;
        // Kísérlet: választás végrehajtása, állapot frissítése
        backtrack(choices, state + choice, n, res, len);
        // Visszalépés
    }
}

/* Lépcsőmászás: visszalépés */
int climbingStairsBacktrack(int n) {
    int choices[2] = {1, 2}; // Egyszerre 1 vagy 2 lépcsőfokot lehet mászni
    int state = 0;           // Indulás a 0. lépcsőfokról
    int *res = (int *)malloc(sizeof(int));
    *res = 0; // res[0] tárolja a megoldások számát
    int len = sizeof(choices) / sizeof(int);
    backtrack(choices, state, n, res, len);
    int result = *res;
    free(res);
    return result;
}

/* Vezérlő kód */
int main() {
    int n = 9;

    int res = climbingStairsBacktrack(n);
    printf("Climbing %d stairs has %d solutions\n", n, res);

    return 0;
}

File: hu/codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js
/**
 * File: climbing_stairs_backtrack.js
 * Created Time: 2023-07-26
 * Author: yuan0221 (yl1452491917@gmail.com)
 */

/* Visszalépés */
function backtrack(choices, state, n, res) {
    // Ha felértünk az n-edik lépcsőfokra, a megoldások számát 1-gyel növeljük
    if (state === n) res.set(0, res.get(0) + 1);
    // Bejárjuk az összes választási lehetőséget
    for (const choice of choices) {
        // Metszés: nem engedünk meg n-nél tovább menni
        if (state + choice > n) continue;
        // Kísérlet: választás megtétele, állapot frissítése
        backtrack(choices, state + choice, n, res);
        // Visszalépés
    }
}

/* Lépcsőmászás: Visszalépés */
function climbingStairsBacktrack(n) {
    const choices = [1, 2]; // Választhatunk 1 vagy 2 lépcsőfokot
    const state = 0; // A 0. lépcsőfoktól kezdjük
    const res = new Map();
    res.set(0, 0); // res[0] értékével rögzítjük a megoldások számát
    backtrack(choices, state, n, res);
    return res.get(0);
}

/* Tesztkód */
const n = 9;
const res = climbingStairsBacktrack(n);
console.log(`Climbing ${n} stairs has ${res} solutions`);
